zarr/hierarchy.py

```diff
@@ -0,0 +1,850 @@
+                arrays_line = arrays_line[:77] + '...'
+            r += arrays_line
+        group_keys = list(self.group_keys())
+        if group_keys:
+            groups_line = '\n  groups: %s; %s' % \
+                (len(group_keys), ', '.join(group_keys))
+            if len(groups_line) > 80:
+                groups_line = groups_line[:77] + '...'
+            r += groups_line
+        r += '\n  store: %s.%s' % (type(self._store).__module__,
+                                   type(self._store).__name__)
+        return r
+
+    def _item_path(self, item):
+        if item and item[0] == '/':
+            # absolute path
+            path = normalize_storage_path(item)
+        else:
+            # relative path
+            path = normalize_storage_path(item)
+            if self.path:
+                path = self.path + '/' + path
+        return path
+
+    def __contains__(self, item):
+        """Test for group membership.
+
+        Examples
+        --------
+        >>> import zarr
+        >>> g1 = zarr.group()
+        >>> g2 = g1.create_group('foo')
+        >>> d1 = g1.create_dataset('bar', shape=100, chunks=10)
+        >>> 'foo' in g1
+        True
+        >>> 'bar' in g1
+        True
+        >>> 'baz' in g1
+        False
+
+        """
+        path = self._item_path(item)
+        if contains_array(self.store, path):
+            return True
+        elif contains_group(self.store, path):
+            return True
+        else:
+            return False
+
+    def __getitem__(self, item):
+        """Obtain a group member.
+
+        Parameters
+        ----------
+        item : string
+            Member name or path.
+
+        Examples
+        --------
+        >>> import zarr
+        >>> g1 = zarr.group()
+        >>> d1 = g1.create_dataset('foo/bar/baz', shape=100, chunks=10)
+        >>> g1['foo']
+        zarr.hierarchy.Group(/foo, 1)
+          groups: 1; bar
+          store: zarr.storage.DictStore
+        >>> g1['foo/bar']
+        zarr.hierarchy.Group(/foo/bar, 1)
+          arrays: 1; baz
+          store: zarr.storage.DictStore
+        >>> g1['foo/bar/baz']
+        zarr.core.Array(/foo/bar/baz, (100,), float64, chunks=(10,), order=C)
+          compression: blosc; compression_opts: {'clevel': 5, 'cname': 'lz4', 'shuffle': 1}
+          nbytes: 800; nbytes_stored: 283; ratio: 2.8; initialized: 0/10
+          store: zarr.storage.DictStore
+
+        """  # flake8: noqa
+        path = self._item_path(item)
+        if contains_array(self.store, path):
+            return Array(self.store, readonly=self.readonly, path=path)
+        elif contains_group(self.store, path):
+            return Group(self.store, readonly=self.readonly, path=path)
+        else:
+            raise KeyError(item)
+
+    def __setitem__(self, key, value):
+        """Not implemented."""
+        raise NotImplementedError()
+
+    def keys(self):
+        """Return an iterator over member names.
```

Comment:
If you define `__iter__` instead and make this in a `collections.Mapping`, Python will give you reentrant iterator versions of `keys()`, `values()` and `items()` automatically.